Add support for succinct hash bin delegation

MVrachev opened the pull request:
**Note this pr is dependant on #33**

With this pr I add support for succinct hash bin delegation.

The pr introduces two commands:
1. `tufrepo edit X add-delegation --succinct <options>`  edits targets role X, making the delegation into a succinct one
2. `tufrepo init-succinct-roles X` reads the delegation and initializes all the delegated metadata files

There are additional commits that are aiming to simplify or improve the codebase.

MVrachev commented:
The tests are failing due to `python-tuf` not yet released a new version which will add support for succinct hash bin delegation.

jku commented on tufrepo/cli.py:
```diff
@@ -115,6 +115,39 @@ def snapshot(ctx: Context):
+            helpers.add_key(targets, role, None, new_key.public)
+
+        for bin_name in targets.delegations.succinct_roles.get_roles():
+            # Store all generated keys for delegations for each of the bins.
+            for key in full_keys_info:
+                ctx.obj.keyring.store_key(bin_name, key)
```
the way privkeys.json size explodes here leads me to think the format is not a great design... but the whole privkeys.json is a hack in any case: succinct delegations are meant for environment keys anyway

No need to do anything about this here.

jku commented:
Looks pretty good. I'll leave a naming comment already:

> 1. `tufrepo edit X add-delegation --succinct <options>`  edits targets role X, making the delegation into a succinct one

this does not seem to be true... you actually have `tufrepo edit X add-succinct-delegation ...` and a separate `tufrepo edit X add-delegation` for paths and hash_prefixes. I think I still prefer a single command with options to switch between the three modes (paths/hash_prefixes/succinct) but I'm willing to be persuaded.

Also, I think there's no good default name prefix: there's nothing special about "bin" and adding two succinct delegations with default name to the same repo would or should fail anyway. It seems reasonable to me to make both number and name-prefix required when adding a succinct delegation.

jku reviewed (commented):
I mentioned two CLI user experience issues already, I think otherwise I just have some comments/nitpicks: looks good overall!

I think it would be good to change requirements.txt so it uses a git commit for python-tuf -- then the tests should work on CI and I can also easily tests this locally?

jku commented:
WRT succinct UI for tufrepo: according to click docs this should give you a `Tuple[str, int]`:
`@click.option('--succinct', type=(str, int))`
that should allow `tufrepo edit X add-delegation --succinct bin 32`: not amazing UI but not horrible

jku commented on tufrepo/cli.py:
```diff
@@ -98,9 +111,44 @@ def verify(ctx: Context, root_hash: Optional[str] = None):
+            )
+
+        full_keys_info: List[PrivateKey] = []
+        for _ in range(targets.delegations.succinct_roles.threshold):
+            new_key: PrivateKey = ctx.obj.keyring.generate_key()
+            full_keys_info.append(new_key)
+            helpers.add_key(targets, role, None, new_key.public)
```
this is a bit magical -- uses threshold info from the delegator but does not use any possibly existing keys from delegator, instead creates new ones.

I don't have a great alternative ideas yet so I think this is fine for now. I think a better solution would first require some smarter solution for the private key storage for succinct keys

MVrachev commented on tufrepo/cli.py:
```diff
@@ -255,6 +261,35 @@ def add_delegation(
+):
+    """Add succinct hash bin delegation to a ROLE representing NUMBER of bins.
+
+    NUMBER must be a power of 2, otherwise, it will be casted down.
```
I think raising an error will be the right solution here.

MVrachev commented on tufrepo/cli.py:
```diff
@@ -248,18 +248,23 @@ def add_delegation(
 
         role = DelegatedRole(delegate, [], 1, terminating, _paths, _prefixes)
-        targets.delegations.roles[role.name] = role
+        assert tar.delegations.roles is not None
```
It was suggested by an extension I have.
It says: `Object of type \"None\" is not subscriptable`.

Usually, it gives good suggestions, but I am not sure about this one.
Here as we know that if a `targets` didn't had a delegation one will be created or if did have with succinct hash delegation it will be overwritten. 
We also know that one of `roles` or `succinct_roles` should be set in  a `Delegation` object meaning this is a little extra check.
Will remove it.

jku commented:
```
@edit.command()
@click.pass_context
@click.option("--terminating/--non-terminating", default=False)
@click.option("--path", "paths", multiple=True)
@click.option("--hash-prefix", "hash_prefixes", multiple=True)
@click.option("--succinct", "bin_count", type=int)
@click.argument("delegate")
def add_delegation(
    ctx: Context,
    terminating: bool,
    paths: Tuple[str],
    hash_prefixes: Tuple[str],
    bin_count: int,
    delegate: str,
):
```

still mulling this over... the above could work: the only surprising thing here is that in the succinct case "delegate" is the name prefix for the delegate, and I think that makes some sense:

```
tufrepo edit X add-delegate --path "a/b" --path "c/d" delegaterole
tufrepo edit X add-delegate --hash-prefix "00" --hash-prefix "01" delegaterole
tufrepo edit X add-delegate --succinct 32 bin
```
these look reasonable to me...

MVrachev commented:
@jku I followed your suggestion and made `add-delegation`, so it can be used both for adding new delegated role and adding succinct hash bin delegation.
Additionally, I noticed some small stuff that needed polishing when testing a command that we expect to fail.
See https://github.com/vmware-labs/repository-editor-for-tuf/pull/34/commits/f6a48d156fbca3dab20803254f8d13776fa67814

jku commented on tufrepo/cli.py:
```diff
@@ -230,28 +271,73 @@ def remove_target(ctx: Context, target_path: str):
 @click.option("--hash-prefix", "hash_prefixes", multiple=True)
+@click.option("--succinct", "bin_amount", type=int)
+@click.argument("delegate")
 def add_delegation(
```
Also raising value errors for wrong CLI input looks really hostile: **if** we check input, let's print good error messages and not stack traces.

Click has a way to do this: ClickExceptions will be handled and printed neatly (there are examples in verifier at least)

jku commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 securesystemslib[pynacl]
-tuf==1.1.0
+# Build from source the latest changes in the develop branch of python-tuf.
+git+https://github.com/theupdateframework/python-tuf.git@develop
```
I would prefer a specific hash please

jku replied:
Added a commit for this already

jku commented:
> Additionally, I noticed some small stuff that needed polishing when testing a command that we expect to fail.

I don't fully agree with that change... We can merge this but for future:
* The reason the current test code did not work for you is that CLI outputs a stack trace if user makes a mistake. That is just bad UI: I don't think we should write tests that verify that we output bad UI.
* I don't mind changing how _run() checks the output but now stdout and stderr are handled completely differently...

I would suggest to use e.g. ClickException to get reasonable output on input validation failure instead (see examples in e.g. verifier).

jku reviewed (approved):
I had comments about the input validation and how its tested but... let's get this merged and see about those later: I'd really like to start testing this in tuf-demo etc.

Let's not use a moving git branch as a requirement... I can fix that before I merge

MVrachev commented:
I am going to propose a pr addressing all of those comments.

I have a question you said:
> there is a check for powers of two that is broken

In which case is broken?

jku commented:
> > there is a check for powers of two that is broken
> 
> In which case is broken?

Most cases, (i % 2) is just not a check for powers of two.

MVrachev commented:
Ohh... myyy bad. You are correct.
I played myself my testing with an odd number - 3 instead of testing with an even number that it's not a power of 2.
